provider/pkg/gen/go-templates/apiextensions/customResource.tmpl

```diff
@@ -0,0 +1,131 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// *** WARNING: this file was generated by pulumigen. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+package apiextensions
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pulumi/pulumi-kubernetes/sdk/v2/go/kubernetes"
+	metav1 "github.com/pulumi/pulumi-kubernetes/sdk/v2/go/kubernetes/meta/v1"
+	"github.com/pulumi/pulumi/sdk/v2/go/pulumi"
+)
+
+// CustomResource represents a resource definition we'd use to create an instance of a
+// Kubernetes CustomResourceDefinition (CRD). For example, the CoreOS Prometheus operator
+// exposes a CRD `monitoring.coreos.com/ServiceMonitor`; to create a `ServiceMonitor`, we'd
+// pass a `CustomResourceArgs` containing the `ServiceMonitor` definition to
+// `apiextensions.CustomResource`.
+//
+// NOTE: This type is abstract. You need to inherit from it and define all the properties of
+// a specific custom resource in the derived class.
+type CustomResource struct {
+	pulumi.CustomResourceState
+
+	// APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+	ApiVersion pulumi.StringOutput `pulumi:"apiVersion"`
+	// Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	Kind pulumi.StringOutput `pulumi:"kind"`
+	// Standard object metadata; More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata.
+	Metadata metav1.ObjectMetaPtrOutput `pulumi:"metadata"`
+	// Untyped map that holds any user-defined fields.
+	OtherFields pulumi.MapOutput `pulumi:""`
+}
+
+// NewCustomResource registers a new resource with the given unique name, arguments, and options.
+func NewCustomResource(ctx *pulumi.Context,
+	name string, args *CustomResourceArgs, opts ...pulumi.ResourceOption) (*CustomResource, error) {
+	if args == nil {
+		args = &CustomResourceArgs{}
+	}
+
+	untyped := args.OtherFields
+	untyped["apiVersion"] = args.ApiVersion
+	untyped["kind"] = args.Kind
+	untyped["metadata"] = args.Metadata
```

Comment:
Will this mutate the dict passed in by the user?  Should we copy fields into a new dict here?

Reply:
Possibly so. I’ll copy to a new map